[Archit Goswami] iP

ararchch opened the pull request:
# IP Update: 

>Introducing Yapchit, your new task assistant.

This PR introduces several enhancements the existing Duke implementation, improving the number of operations handled as well as introducing error handling. Below is a summary of the key changes:

- **Implemented new operations:** Implemented new task management functions including create, mark, unmark, delete.
- **Improved code quality:** Introduced OOP features such as new classes to improve code quality.
- **Included JavaDocs:** Included JavaDocs for key methods
- **Exception handling:** Introduced exception handling to counteract adverse inputs.
- **Tests:** Included tests for some methods.

```java
// Example of code block from Yapchit with JavaDocs, OOP features, exception handling, and operations. 
/**
     * Primary function is to accept an input and operation and redirect it to appropriate handler function
     * for handling.
     *
     * @param input user input
     * @param op parsed operation from parser
     * @param tasks the full tasklist
     * @param ui the ui to interact with user
     * @param parser the parser to perform additional parsing
     * @throws YapchitException if there is conflict in terms of user input and operation to perform
     */
    public void handleOperation(
            String input, Yapchit.Operations op, TaskList tasks, Ui ui, Parser parser) throws YapchitException {
        String[] parts = parser.parseInputParts(input);

        switch (op) {
        case LIST:
            handleList(parts, tasks, ui);
            break;

        case MARK:
            handleMark(parts, tasks, ui);
            break;

        case UNMARK:
            handleUnmark(parts, tasks, ui);
            break;

        case DELETE:
            handleDelete(parts, tasks, ui);
            break;

        case FIND:
            handleFind(parts, tasks, ui);
            break;

        case DEADLINE:
            handleDeadline(input, true, tasks, ui, parser);
            break;

        case EVENT:
            handleEvent(input, true, tasks, ui);
            break;

        case TODO:
            handleTodo(input, true, tasks, ui);
            break;

        default:
            throw new InvalidKeywordException("You have entered an invalid keyword. " +
                    "Valid keywords are ['mark', 'unmark', 'deadline', 'todo', 'event', 'bye', 'list', 'delete']");
        }
    }
```

## How to clone and run the code:
1. Clone the repository to your local drive.
2. `cd` into the `yapchit` directory
3. Compile and run the Yapchit.java class using your IDE or via command line.
4. Interact with Yapchit using your terminal. Here is an example of what you can say: `todo swimming` or `mark 1`


### Features intended to be implemented
- [x] Accept different types of tasks.
- [x] Enable persistence of tasks across Yapchit restarts.
- [ ] Implement a UI.
- [ ] ~~Use chatGPT~~

🚀 **While not complete, Yapchit is still a big step towards a fully functional task manager.**

For more details and latest updates visit the _yapchit_ repo [here](https://github.com/ararchch/ip)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ararchch committed:
Renamed bot, created intro + exit message

ararchch committed:
Renamed class, implemented echo function

ararchch committed:
Refactored main into methods, implemented list functionality

ararchch committed:
Implemented mark, unmark functionality and limited error handling

ararchch committed:
Fixed bug, implemented functionality for handling multiple tasks

Fixed bug from level 3 where you could not enter jobs that were longer than one word.

Implemented task functionality necessary for level 4

ararchch committed:
Implemented test case, minor changes to appearance

ararchch committed:
Created Yapchit exception classes, implemented exception handling

ararchch committed:
Implemented delete, refactored if statements to switch statements

ararchch committed:
Included enums, minor other fixes

ararchch committed:
refactored code into shorter methods

ararchch committed:
minor fixes

ararchch committed:
minor fixes #2

ararchch committed:
created function to handle import from file

ararchch committed:
created fileUpdate functionality

ararchch committed:
minor fixes

ararchch committed:
Implement date object functionality in deadline

Implemented functionality to handle dates provided with deadlines as a date object rather than as Strings, improving flexibility.

ararchch committed:
Merge branch 'branch-Level-8'

ararchch committed:
Split implementation into multiple classes

Created an initial split of Yapchit class into multiple classes. Finished implementation of TaskList, Handler, Parser classes.

ararchch committed:
Complete implementation of more OOP

Finished implementation of more OOP. Split Yapchit class into 5 other classes, each handling a different function.

- Yapchit: Main Yapchit class, encapsulates the entire bot.
- Handler: Contains primary functionality to handle operations and requests. Acts as the 'brains'
- Parser: Parses input
- Ui: Deals with accepting user input and various forms of output (printing) on the screen
- TaskList: Encapsulates the list of tasks and performs list operations.
- Storage: Deals with the storage and retrieval of tasks to/from the computer's memory.

ararchch committed:
Merge branch 'branch-A-MoreOOP'

ararchch committed:
Refactor code into Yapchit package.

Created a Yapchit package to encompass codebase.

ararchch committed:
Merge branch 'branch-A-Packages'

ararchch committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

ararchch committed:
Create HandlerTest file along with testcases

Created a new class called HandlerTest that contains comprehensive testcases to test the handleDeadline method in the handler class.

ararchch committed:
Create ParserTest file along with testcases

Created a new class called PaserTest that contains comprehensive testcases to test the parseInput method in the parser class.

ararchch committed:
Merge branch 'branch-A-JUnit'

ararchch committed:
Create jar file

ararchch committed:
Merge branch 'branch-A-Jar'

ararchch committed:
Implement find functionality

Implemented the ability to find tasks based on a search term.
Modified printList feature to accommodate this functionality.

ararchch committed:
Improve coding standard

Made minor fixes throughout the program to improve coding standard

ararchch committed:
Add JavaDocs

Added JavaDocs for improved readability of code.

ararchch committed:
Merge branch 'branch-Level-9'

ararchch committed:
Merge remote-tracking branch 'origin/branch-A-JavaDoc'

# Conflicts:
#	src/main/data/dataStore.txt
#	src/main/java/Yapchit/Handler.java
#	src/main/java/Yapchit/TaskList.java
#	src/main/java/Yapchit/Ui.java

ararchch committed:
Minor fixes

ararchch committed:
Change package names

ararchch committed:
Fix imports in test package

ararchch committed:
Resolve merge conflict

ararchch committed:
Resolve merge conflict #2

Completed merge of 3 branches, resolved all conflicts.

ararchch committed:
Small bug fixes

ReflectiveObsidian commented on src/main/java/Yapchit/Tasks/ToDo.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * Prints the ToDo details onto the screen.
+     *
+     * @return
```
The return description of this javadoc is missing

ararchch replied:
ah yes, thanks for pointing it out

ReflectiveObsidian commented on src/main/java/yapchit/TaskList.java:
```diff
@@ -0,0 +1,83 @@
+
+    /**
+     * adds task to list.
+     * @param t task to add.
```
Should there be an empty line between description and parameter section?

ararchch replied:
yup! that would improve readability

ReflectiveObsidian commented on src/main/java/yapchit/tasks/Task.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    /**
+     * setter function to set the tag of the class based on input.
```
Suggestion to follow the java coding standard more closely:
```suggestion
     * Sets the tag of the class based on input.
```

ReflectiveObsidian commented on src/test/java/yapchit/HandlerTest.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Test
+    void testHandleDeadlineSuccess() throws YapchitException {
```
Suggestion to name test cases according to the java coding standard:
```suggestion
    void handleDeadline_variedSpaces_success() throws YapchitException {
```

ReflectiveObsidian reviewed (commented):
Looks good. Some suggestions for coding standard.

ararchch committed:
Create basic yapchit interface

Interface created without link to backend

ararchch committed:
Connect ui with backend logic

ararchch committed:
Update handling of missing file error

ararchch committed:
Update imports to remove wildcard

ararchch committed:
Merge branch 'branch-Level-10'

ararchch commented:
Thanks for your inputs! Will improve on them moving forward.

ararchch committed:
Fix folder name issue step 1

ararchch committed:
Fix folder name issue step 2

ararchch committed:
Update build.gradle

ararchch committed:
Include assertions at key parts of the code

Include assertions at key parts of the code

The current code lacks assertions at relevant spots.

Assertions are important to ensure that bugs in the code are caught
before they cause worse problems.

Including assertions at relevant spots in the code, primarily the
Handler class and the Yapchitbackend class ensures:
* safe transfer of data between different components of Yapchit
* more bugs that can cause errors down the line are caught early

As the Handler class and Yapchitbackend primarily facilitate
this communication between components it is important to
at least include assertions there.

ararchch committed:
Merge pull request #1 from ararchch/branch-A-Assertions

Include assertions at key parts of the code

ararchch committed:
Revert "Merge pull request #1 from ararchch/branch-A-Assertions"

This reverts commit 9b12cec69ec44e8feac5bceceb4aa9d676cf556b, reversing
changes made to 997f76effe36fe518adc4f67634edeae246e64f8.

ararchch committed:
Improve code quality

Code Quality Improvement: Enhance readability, reduce unsafe practices

Yapchit, Handler, Yapchitbackend classes (primarily) have code that
it not compliant with code quality rules, such as functions with common
behaviours, recycled parameters, and large methods etc.

These aspects of the code hinder readability while also making the
code more bug prone.

Key approaches taken to solve the issues:
* Wherever possible and reasonably impactful, common behaviour was
  abstracted out into a separate method.
* New variables were introduced to eliminate recycling of parameters
* Large methods were split into smaller ones.
* Variables were brought closer to the scope in which they are used
Most of the improvements were in the Handler class and the Yapchit
class.

These methods were adopted because they allowed the code to fit
the code quality requirements as outline by the textbook. They are
also typical practice when it comes to addressing similar problems
(i.e. abstraction etc.)

ararchch committed:
Merge pull request #2 from ararchch/branch-A-CodeQuality

Improve code quality

ararchch committed:
Include more assertions

Include more asserations

While current code has some assertions
more help to ensure more bugs are caught.

Assertions were inserted at prominent locations
in yapchitbackend, storage and other classes where
there were sensitive processes taking place and
interchange of data.

Such places are where bugs can occur, especially if
there is transfer between two classes. Often times
when writing the code, the two classes were written
in different times of the day or different days, hence
potential for error and bugs is higher.

ararchch committed:
Fix commit issues

This fixes issues relating to commits

I originally created a commit with assertions. However, I
merged the PR early, so I undid the changes in the
commit. However, when I fixed it and went to create
the PR at the correct time, the changes I had initially
merger were not reflected in the PR, but were on the
branch.

This commit fixes that issue by editing each of the
original assertions slightly

This is so that git recognises it as a new change and
hence registers them in the new PR.

ararchch committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/yapchit/yapchitbackend/Handler.java
#	src/main/java/yapchit/yapchitbackend/Storage.java
#	src/main/java/yapchit/yapchitbackend/YapchitBackend.java
#	src/main/java/yapchit/yapchitui/Yapchit.java

ararchch committed:
Merge pull request #3 from ararchch/branch-A-Assertions

Branch a assertions

ararchch committed:
Implement task update functionality

Implement ability to update tasks

Currently tasks cannot be edited once created

This limits functionality of the bot as editing
tasks is a natural thing to do.

Including and handling an update keyword
allows people to use this function.

The keyword is necessary to inform that this is
a new function. It also enables seamless
incorporation along with the existing commands

Update format is:
update [task index] [new name] [/param] [new param]
If a parameter or the name is not being updated,
you may use the '*' to notify the program that
you wish to keep the existing value.

Consequently, '*' is now a reserved character and cannot
be used like a typical one. This decision limits inputs
slightly but the decision was made because it is unlikely
that a user will enter just a '*' as a name or description
to any task.

Examples of update commands:
 - update 1 eat /by 2020-11-11
 - update 4 say hello
 - update 2 * /from * /to tomorrow

ararchch committed:
Merge pull request #4 from ararchch/branch-BCD-Extension

Implement task update functionality

ararchch committed:
Minor update to ui

ararchch committed:
Organise javafx code

ararchch committed:
Update readme.md

ararchch committed:
Minor fixes

ararchch committed:
Update ui image

ararchch committed:
Delete Ui.png

ararchch committed:
Create Ui.png

ararchch committed:
More minor fixes
